warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,95 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pyramid.authorization import Authenticated, Everyone
+from pyramid.interfaces import ISecurityPolicy
+from zope.interface import implementer
+
+from warehouse.accounts.interfaces import IUserService
+
+
+class SecurityPolicy:
+    def __init__(self, callback=None):
+        self._callback = callback
+
+    def identity(self, request):
+        login_service = request.find_service(IUserService, context=None)
+        user = login_service.get_user(self.unauthenticated_userid(request))
+        if user is None:
+            return None
+
+        principals = []
+        if self._callback is not None:
+            principals = self._callback(user.id, request)
+            if principals is None:
+                return None
+
+        return {"entity": user, "principals": principals}
+
+    def permits(self, request, context, permission):
+        return NotImplemented
+
+
+@implementer(ISecurityPolicy)
+class MultiSecurityPolicy:
+    """
+    A wrapper for multiple Pyramid 2.0-style "security policies", which replace
+    Pyramid 1.0's separate AuthN and AuthZ APIs.
+
+    Security policies are checked in the order provided during initialization,
+    with the following semantics:
+
+    * `identity`: Selected from the first policy to return non-`None`
+    * `authenticated_userid`: Selected from the request identity, if present
+    * `forget`: Combined from all policies
+    * `remember`: Combined from all policies
+    * `permits`: Uses the AuthZ policy passed during initialization
+
+    These semantics mostly mirror those of `pyramid-multiauth`.
+    """
+
+    def __init__(self, policies, authz):
+        self._policies = policies
+        self._authz = authz
+
+    def identity(self, request):
+        for policy in self._policies:
+            if ident := policy.identity(request):
+                return ident
+
+        return None
+
+    def authenticated_userid(self, request):
+        if request.identity:
+            return str(request.identity["entity"].id)
+        return None
+
+    def forget(self, request, **kw):
+        headers = []
+        for policy in self._policies:
+            headers.extend(policy.forget(request, **kw))
+        return headers
+
+    def remember(self, request, userid, **kw):
+        headers = []
+        for policy in self._policies:
+            headers.extend(policy.remember(request, userid, **kw))
+        return headers
+
+    def permits(self, request, context, permission):
+        identity = request.identity
+        principals = [Everyone]
```

Comment:
Flagging this for review: I cargo-culted this from the Pyramid cookbook. I don't _think_ it causes access issues since `Everyone` means "everyone, all the time" and as such it's sort of a no-op check against most routes, but I'm not certain.

Reply:
None that I can think of, they should be functionally equivalent. Will drop.

Reply:
Not necessarily asking to drop it, just don't understand what it's purpose is. Maybe we should be doing it?

Reply:
If I understand how Pyramid does ACLs correctly, it's essentially the "no-op" principal -- every identity belongs to `Everyone`, including "no-one", so there's no difference between it being present and not.

That's per the docs:

> [](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/authorization.html#pyramid.authorization.Everyone)Everyone
The special principal id named Everyone. This principal id is granted to all requests. Its actual value is the string 'system.Everyone'.

Reply:
Removed it and nothing broke.